src/llama_stack/core/routers/__init__.py

```diff
@@ -45,6 +45,18 @@ async def get_routing_table_impl(
         raise ValueError(f"API {api.value} not found in router map")
 
     impl = api_to_tables[api.value](impls_by_provider_id, dist_registry, policy)
+
+    # Apply tracing to routing table if any base class has __marked_for_tracing__ marker
+    # (Tracing will be no-op if telemetry is disabled)
+    traced_classes = [
+        base for base in reversed(impl.__class__.__mro__) if getattr(base, "__marked_for_tracing__", False)
+    ]
+    if traced_classes:
+        from llama_stack.core.telemetry.trace_protocol import trace_protocol
+
+        for cls in traced_classes:
+            trace_protocol(cls)
```

Comment:
I wonder if you could do this when you do an `instantiate_provider()` call -- that's the only place an impl is created. would that be enough to cover all the cases?